hare3/algorithm.go

```diff
@@ -0,0 +1,478 @@
+// actions of message decoding and signature verification and key grading,
+// leaving this handler to handle the decoded message inputs.
+type Handler struct {
+	gg  GradedGossiper
+	tgg ThresholdGradedGossiper
+	gc  Gradecaster
+	rp  *RoundProvider
+	mu  *sync.Mutex
+}
+
+func NewHandler(gg GradedGossiper, tgg ThresholdGradedGossiper, gc Gradecaster, rp *RoundProvider, protocolMu *sync.Mutex) *Handler {
+	return &Handler{
+		gg:  gg,
+		tgg: tgg,
+		gc:  gc,
+		rp:  rp,
+		mu:  protocolMu,
+	}
+}
+
+// HandleMsg handles an incoming message, it returns a boolean indicating
+// whether the message should be regossipped to peers.
+func (h *Handler) HandleMsg(hash types.Hash20, vk []byte, round AbsRound, values []types.Hash20) (bool, *types.Hash20) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	r := AbsRound(round)
+	var g uint8
+	switch r.Type() {
+	case Propose:
+		g = gradeKey3(vk)
+	default:
+		g = gradeKey5(vk)
+	}
+	id := hashBytes(vk)
+
+	var equivocationHash *types.Hash20
+	// Nil values signifies a message from a known malicious actor, however we
+	// still need to forward the first such message per identity to the network
+	// though so that they also insert it into their protocols.
+	result, eHash := h.gg.ReceiveMsg(hash, id, values, r)
+	switch result {
+	case DropMessage:
+		// Indicates prior equivocation, drop the message.
+		return false, nil
+	case SendEquivocationProof:
+		// Indicates new instance of equivocation notify gradecast or threshold
+		// gossip with nil values.
+		// Set the equivocationHash so that the caller can look up the message.
+		equivocationHash = eHash
+	}
+
+	curr := h.rp.CurrentRound()
+	// In the case of early messages we just set the curr round to be equal to
+	// the message round. I.E. we assume they were delivered perfectly on time.
+	if curr < round {
+		curr = round
+	}
+
+	// Pass results to gradecast or threshold gossip.
+	// Only proposals are gradecasted, all other message types are passed to
+	// threshold gossip.
+	if r.Type() == Propose {
+		// Send to gradecast
+		h.gc.ReceiveMsg(id, values, r, curr, g)
+	} else {
+		// Pass result to threshold gossip
+		h.tgg.ReceiveMsg(id, values, r, curr, g)
+	}
+	return true, equivocationHash
+}
+
+func (h *Handler) Protocol(lc LeaderChecker, si []types.Hash20) *Protocol {
+	return NewProtocol(h.tgg, h.gc, lc, h.mu, si)
+}
+
+type Protocol struct {
+	// round encodes both iteration and round in a single value.
+	round AbsRound
+	// Each index "i" holds an indication of whether iteration "i" is hard locked
+	hardLocked []bool
+	// Each index "i" holds the locked value for iteration "i"
+	Li []*types.Hash20
+	// Each index "i" holds values considered valid up to round "i+1"
+	Vi [][]types.Hash20
+	// Each index "i" holds a map of sets of values from valid proposals
+	// indexed by their hash received in iteration "i"
+	Ti  []map[types.Hash20][]types.Hash20
+	Si  []types.Hash20
+	tgg ThresholdGradedGossiper
+	gc  Gradecaster
+	lc  LeaderChecker
+	mu  *sync.Mutex
```

Comment:
please put it on top of stuff that is synchronized in final version

Reply:
Do you mean the following ?

```
mu *sync.Mutex
x     someType // synchronised by mu

y     someOtherType // not synchronised by mu
```
Is this a convention that we have, or in general a convention? I've never heard of this before.

Reply:
yes like that, it is a convention atleast in spacemesh codebase, and i definitely stumbled on it before, though i will need to find sources if you need them

another approach that i like is to nest sychronized fields after mutex, it is general practice in cockroachdb codebase

```
mu struct {
    sync.Mutex
    x someType
}
```

Reply:
The nesting is nice, more explicit and less likely to get mangled by future updates

Reply:
I'd be in favour of nesting, unless you'd rather stick with the line locality approach?